### ISSUE TITLE
log_model throws assertion error when creating tmp dir. Running in a pyspark process

### ISSUE THREAD

**Issue by sstrac** [mlflow/mlflow#12157]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

MLflow version: 2.12.1

### System information

System information: Linux #1 SMP Sat Nov 4 10:20:54 UTC 2023
Python version: 3.10.12

### Describe the problem

When trying to log model with mlflow in a pyspark pipeline we get an assertion error, mlflow is calling TempDir and trying to assert current working directory exists which fails. Observations:
- The error occurs on around 80% of runs, other times it passes with no code change between runs.
- The failed run prints logs from the main function twice i.e. context {...}, successful runs print it once.
- The pipeline works in a databricks notebook with the same cluster memory and versions.
- Even in failed runs the model does end up getting logged, parameters (the next line of code) are not logged and the spark job still fails.

### Tracking information

```shell
System information: Linux #1 SMP Sat Nov 4 10:20:54 UTC 2023
Python version: 3.10.12
MLflow version: 2.12.1
MLflow module location: /databricks/python/lib/python3.10/site-packages/mlflow/__init__.py
Tracking URI: databricks
Registry URI: databricks
Databricks runtime version: 14.3.x-cpu-ml-scala2.12
MLflow environment variables: 
  MLFLOW_CONDA_HOME: /databricks/conda
  MLFLOW_DEPLOYMENTS_TARGET: databricks
  MLFLOW_GATEWAY_URI: databricks
  MLFLOW_PYTHON_EXECUTABLE: /databricks/spark/scripts/mlflow_python.sh
  MLFLOW_TRACKING_URI: databricks
MLflow dependencies: 
  Flask: 3.0.3
  Jinja2: 3.1.4
  aiohttp: 3.9.1
  alembic: 1.13.1
  azure-storage-file-datalake: 12.14.0
  boto3: 1.24.28
  botocore: 1.27.96
  click: 8.1.7
  cloudpickle: 3.0.0
  docker: 7.0.0
  entrypoints: 0.4
  gitpython: 3.1.43
  google-cloud-storage: 2.11.0
  graphene: 3.3
  gunicorn: 21.2.0
  importlib-metadata: 7.1.0
  markdown: 3.6
  matplotlib: 3.8.4
  numpy: 1.26.4
  packaging: 24.0
  pandas: 1.5.3
  protobuf: 5.26.1
  pyarrow: 15.0.2
  pydantic: 1.10.6
  pytz: 2024.1
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.31.0
  scikit-learn: 1.4.2
  scipy: 1.13.0
  sqlalchemy: 2.0.30
  sqlparse: 0.5.0
  tiktoken: 0.5.2
  virtualenv: 20.16.7
```

### Code to reproduce issue

```

from typing import Any, Dict
import traceback
import mlflow
from pyspark.sql import SparkSession
from retail_sales_data_product.training.model import (
    BuyerPredictionModel
)
from xgboost import XGBClassifier
import os
from mlflow.utils.file_utils import TempDir

def log_model(
    spark: SparkSession,
    context: Dict[str, Any],
    buyer_prediction_model: BuyerPredictionModel,
) -> None:
    try:

        cwd = os.getcwd()
        print("Current Working Directory:", cwd)
        print("Directory Exists:", os.path.exists(cwd))
        with TempDir() as tmp:
            print("Temporary Directory Created:", tmp.path())
            print("Temporary Directory Exists:", os.path.exists(tmp.path()))

        print("Starting model logging")

        if buyer_prediction_model.error_message is None:
            print("Logging precision-recall curve figure")
            mlflow.log_figure(
                buyer_prediction_model.figures["precision_recall_curve"],
                "precision_recall.png",
            )

            print("Logging metrics")
            mlflow.log_metrics(buyer_prediction_model.metrics)

            print("Logging model")
            if isinstance(buyer_prediction_model.classifier, XGBClassifier):
                print("Classifier is XGBClassifier")
                mlflow.xgboost.log_model(
                    buyer_prediction_model.classifier,
                    artifact_path="xgboost-model",
                    signature=buyer_prediction_model.signature,
                )
            else:
                print("Classifier is not XGBClassifier")
                mlflow.sklearn.log_model(
                    buyer_prediction_model.classifier,
                    artifact_path="xgboost-model",
                    signature=buyer_prediction_model.signature,
                )
        else:
            print("Model has error message: %s", buyer_prediction_model.error_message)

        print("print parameter")
        mlflow.log_param("error", buyer_prediction_model.error_message)

    except Exception as e:
        formatted_exception = "".join(traceback.format_exception(e))  # type: ignore
        print("An error occurred during model logging:\n %s", formatted_exception)
        raise

    finally:
        print("Ending MLflow run")
        mlflow.end_run()

```
```
class BuyerPredictionModel:

    def __init__(
        self,
        classifier: Optional[ClassifierMixin],
        signature: Optional[ModelSignature],
        metrics: Optional[Dict[str, float]],
        figures: Optional[Dict[str, Figure]],
        training_data: Optional[TrainTestSample],
        error_message: Optional[str],
    ):
        self.classifier = classifier
        self.signature = signature
        self.metrics = metrics
        self.figures = figures
        self.training_data = training_data
        self.error_message = error_message

    @classmethod
    def with_attributes(
        cls,
        classifier: ClassifierMixin,
        signature: ModelSignature,
        metrics: Dict[str, float],
        figures: Dict[str, Figure],
        training_data: TrainTestSample,
    ):
        return cls(classifier, signature, metrics, figures, training_data, None)

    @classmethod
    def with_error(cls, error_message: str):
        return cls(None, None, None, None, None, error_message)
```
```
@main.command()
@click.option("--run-date", required=True, type=str)
@click.option(
    "--dummy",
    is_flag=True,
    show_default=True,
    default=False,
    help="Train a dummy model",
)
def training(run_date: str, dummy: bool) -> None:
    mlflow.doctor()
    spark, context = _provide_prerequisites()

    context["model_config"] = BUYER_PREDICTION_CONFIG
    context["run_date"] = datetime.strptime(run_date, "%Y-%m-%d")
    context["dates"] = calculate_training_data_dates(
        context["run_date"], context["model_config"]
    )
    context["visitor_advert_engagement_daily_path"] = (
        get_visitor_advert_engagement_daily_path(context)
    )

    print(f"Context: {context}")

    mlflow.set_experiment(experiment_id=get_or_create_training_experiment_id())
    pipeline = training_pipeline(dummy)
    pipeline.run(spark, catalogue=DataCatalogue(), context=context)
```

### Stack trace

Failed Run
```
Context: {...}


Context: {...}

/databricks/python/lib/python3.10/site-packages/mlflow/types/utils.py:394: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


Starting model logging
Logging precision-recall curve figure
Logging metrics
Logging model
Classifier is XGBClassifier


/databricks/python/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")


Uploading artifacts:   0%|          | 0/9 [00:00<?, ?it/s]

An error occurred during model logging:
 %s Traceback (most recent call last):
  File "/databricks/python/lib/python3.10/site-packages/retail_sales_data_product/training/transform.py", line 266, in log_model
    mlflow.xgboost.log_model(
  File "/databricks/python/lib/python3.10/site-packages/mlflow/xgboost/__init__.py", line 270, in log_model
    return Model.log(
  File "/databricks/python/lib/python3.10/site-packages/mlflow/models/model.py", line 620, in log
    with TempDir() as tmp:
  File "/databricks/python/lib/python3.10/site-packages/mlflow/utils/file_utils.py", line 426, in __exit__
    assert os.path.exists(os.getcwd())
AssertionError

Ending MLflow run
```

Successful Run
```
Context: {...}

/databricks/python/lib/python3.10/site-packages/mlflow/types/utils.py:394: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <[https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_](https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values%3E%60_) for more details.
  warnings.warn(


Starting model logging
Logging precision-recall curve figure
Logging metrics
Logging model
Classifier is XGBClassifier


/databricks/python/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")


Uploading artifacts:   0%|          | 0/9 [00:00<?, ?it/s]

print parameter
Ending MLflow run
```

### Other info / logs

log4j spark output tail
```
24/05/27 02:56:49 INFO AsyncEventQueue: Process of event SparkListenerQueryProfileParamsReady(executionId=5, ...) by listener QueryProfileListener took 1.451043753s.
24/05/27 02:56:50 INFO ClusterLoadAvgHelper: Current cluster load: 0, Old Ema: 1.0, New Ema: 0.85 
24/05/27 02:56:53 INFO ClusterLoadAvgHelper: Current cluster load: 0, Old Ema: 0.85, New Ema: 0.0 
24/05/27 02:57:05 INFO DriverCorral: DBFS health check ok
24/05/27 02:57:05 INFO HiveMetaStore: 0: get_database: default
24/05/27 02:57:05 INFO audit: ugi=root	ip=unknown-ip-addr	cmd=get_database: default	
24/05/27 02:57:05 INFO DriverCorral: Metastore health check ok
24/05/27 02:57:23 INFO ProgressReporter$: Reporting partial results for running commands: 1219294969385542347_8159241497435736106_job-763962794618240-run-248556345868937-action-4450748204583789 [597 occurrences]
24/05/27 02:57:48 INFO ProgressReporter$: Removed result fetcher for 1219294969385542347_8159241497435736106_job-763962794618240-run-248556345868937-action-4450748204583789
24/05/27 02:57:52 INFO DriverCorral$: Cleaning the wrapper ReplId-10ebc-e5086-a562c-b (currently in status Idle(ReplId-10ebc-e5086-a562c-b))
24/05/27 02:57:52 INFO DriverCorral$: sending shutdown signal for REPL ReplId-10ebc-e5086-a562c-b
24/05/27 02:57:52 INFO PythonDriverLocalBase$RedirectThread: Python RedirectThread exit
24/05/27 02:57:52 INFO PythonDriverLocalBase$RedirectThread: Python RedirectThread exit
24/05/27 02:57:52 INFO ReplCrashUtils$: python shell exit code: 143; replId: ReplId-10ebc-e5086-a562c-b, pid: 1804
24/05/27 02:57:52 INFO DriverCorral$: sending the interrupt signal for REPL ReplId-10ebc-e5086-a562c-b
24/05/27 02:57:52 INFO DriverCorral$: waiting for localThread to stop for REPL ReplId-10ebc-e5086-a562c-b
24/05/27 02:57:52 INFO DriverCorral$: ReplId-10ebc-e5086-a562c-b successfully discarded
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Executor updated: app-20240527024648-0000/0 is now LOST (worker lost: 10.234.3.197:44119 got disassociated)
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Executor app-20240527024648-0000/0 removed: worker lost: 10.234.3.197:44119 got disassociated
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Master removed worker worker-20240527024626-10.234.3.197-44119: 10.234.3.197:44119 got disassociated
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Worker worker-20240527024626-10.234.3.197-44119 removed: 10.234.3.197:44119 got disassociated
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Executor updated: app-20240527024648-0000/2 is now LOST (worker lost: 10.234.1.133:45013 got disassociated)
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Executor app-20240527024648-0000/2 removed: worker lost: 10.234.1.133:45013 got disassociated
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Master removed worker worker-20240527024641-10.234.1.133-45013: 10.234.1.133:45013 got disassociated
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Worker worker-20240527024641-10.234.1.133-45013 removed: 10.234.1.133:45013 got disassociated
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Executor updated: app-20240527024648-0000/1 is now LOST (worker lost: 10.234.13.5:37503 got disassociated)
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Executor app-20240527024648-0000/1 removed: worker lost: 10.234.13.5:37503 got disassociated
24/05/27 02:58:05 INFO StandaloneAppClient$ClientEndpoint: Master removed worker worker-20240527024623-10.234.13.5-37503: 10.234.13.5:37503 got disassociated
24/05/27 02:58:05 INFO StandaloneSchedulerBackend: Worker worker-20240527024623-10.234.13.5-37503 removed: 10.234.13.5:37503 got disassociated
24/05/27 02:58:06 ERROR TaskSchedulerImpl: Lost executor 0 on 10.234.3.197: worker lost: 10.234.3.197:44119 got disassociated
24/05/27 02:58:06 INFO DAGScheduler: Executor lost: 0 (epoch 30)
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Trying to remove executor 0 from BlockManagerMaster.
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Removing block manager BlockManagerId(0, 10.234.3.197, 37961, None)
24/05/27 02:58:06 INFO BlockManagerMaster: Removed 0 successfully in removeExecutor
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for host: 10.234.3.197 (epoch 30)
24/05/27 02:58:06 INFO TaskSchedulerImpl: Handle removed worker worker-20240527024626-10.234.3.197-44119: 10.234.3.197:44119 got disassociated
24/05/27 02:58:06 ERROR TaskSchedulerImpl: Lost executor 2 on 10.234.1.133: worker lost: 10.234.1.133:45013 got disassociated
24/05/27 02:58:06 INFO TaskSchedulerImpl: Handle removed worker worker-20240527024641-10.234.1.133-45013: 10.234.1.133:45013 got disassociated
24/05/27 02:58:06 ERROR TaskSchedulerImpl: Lost executor 1 on 10.234.13.5: worker lost: 10.234.13.5:37503 got disassociated
24/05/27 02:58:06 INFO TaskSchedulerImpl: Handle removed worker worker-20240527024623-10.234.13.5-37503: 10.234.13.5:37503 got disassociated
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for worker worker-20240527024626-10.234.3.197-44119 on host 10.234.3.197
24/05/27 02:58:06 INFO DAGScheduler: Executor lost: 2 (epoch 32)
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Trying to remove executor 2 from BlockManagerMaster.
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Removing block manager BlockManagerId(2, 10.234.1.133, 45299, None)
24/05/27 02:58:06 INFO BlockManagerMaster: Removed 2 successfully in removeExecutor
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for host: 10.234.1.133 (epoch 32)
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for worker worker-20240527024641-10.234.1.133-45013 on host 10.234.1.133
24/05/27 02:58:06 INFO DAGScheduler: Executor lost: 1 (epoch 34)
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Trying to remove executor 1 from BlockManagerMaster.
24/05/27 02:58:06 INFO BlockManagerMasterEndpoint: Removing block manager BlockManagerId(1, 10.234.13.5, 44623, None)
24/05/27 02:58:06 INFO BlockManagerMaster: Removed 1 successfully in removeExecutor
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for host: 10.234.13.5 (epoch 34)
24/05/27 02:58:06 INFO DAGScheduler: Shuffle files lost for worker worker-20240527024623-10.234.13.5-37503 on host 10.234.13.5
```

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

**Comment by daniellok-db** [comment]
@sstrac what's the output of `print("Current Working Directory:", cwd)` in your `log_model()` code above?

**Comment by ccharlesgb** [comment]
Hey, we have managed to create a more minimal example by running this task:

```
@main.command()
def cwd_mre() -> None:
    spark, context = _provide_prerequisites()
    print(f"Context: {context}")

    while True:
        cwd = os.getcwd()
        print("Current Working Directory:", cwd)
        print("Directory Exists:", os.path.exists(cwd))
        with TempDir() as tmp:
            print("Temporary Directory Created:", tmp.path())
            print("Temporary Directory Exists:", os.path.exists(tmp.path()))
        sleep(30)
```

After 10 minutes this task will fail as the `TempDir` will raise an assertion error that the current working directory does not exist.

After more investigation we have found this issue only happens in Databricks runtime 14 and not in runtime 13. This seems like more of a Databricks issue instead of an mlflow. However maybe mlflow could be made more robust to this problem by not asserting if the current working directory exists if `chdr=False` for the `TempDir` class. If it hasn't changed the current working directory then if the current working directory does not exist then it isn't anything to do with the `TempDir` class and it maybe doesn't make sense for the `TempDir` class to raise the exception in this case.
